File: sydney/Java/JDBC/src/jp/co/ricoh/doquedb/client/PrepareStatement.java
// -*-Mode: Java; tab-width: 4; c-basic-offset: 4;-*-
// vi:set ts=4 sw=4:
//
// PrepareStatement.java -- プリペアステートメントクラス
//
// Copyright (c) 2002, 2003, 2004, 2023 Ricoh Company, Ltd.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

package jp.co.ricoh.doquedb.client;

import jp.co.ricoh.doquedb.common.*;

/**
 * プリペアステートメントクラス
 *
 */
public final class PrepareStatement extends Object
{
	/** データソース */
	private DataSource _dataSource;
	/** データベース名 */
	private String _databaseName;
	/** セッション */
	private Session _session;
	/** プリペアID */
	private int _prepareID;

	/**
	 * 新しくプリペアステートメントオブジェクトを作成する。
	 *
	 * @param dataSource_	データソース
	 * @param databaseName_	データベース名
	 * @param prepareID_	プリペアID
	 */
	public PrepareStatement(DataSource dataSource_,
							String databaseName_,
							int prepareID_)
	{
		super(Object.PREPARE_STATEMENT);
		_dataSource = dataSource_;
		_databaseName = databaseName_;
		_session = null;
		_prepareID = prepareID_;
	}

	/**
	 * 新しくプリペアステートメントオブジェクトを作成する。
	 *
	 * @param session_		セッション
	 * @param prepareID_	プリペアID
	 */
	public PrepareStatement(Session session_,
							int prepareID_)
	{
		super(Object.PREPARE_STATEMENT);
		_dataSource = null;
		_databaseName = null;
		_session = session_;
		_prepareID = prepareID_;
	}

	/**
	 * クローズする。
	 *
	 * @throws	java.io.IOException
	 *			通信関係のエラー
	 */
	public void close()
		throws java.io.IOException
	{
		if (_prepareID != 0)
		{
			try
			{
				if (_session != null)
				{
					_session.erasePrepareStatement(_prepareID);
				}
				else
				{
					Connection clientConnection
						= _dataSource.getClientConnection();
					if (clientConnection != null)
					{
						//プリペアステートメントを削除する
						clientConnection.erasePrepareStatement(
							_databaseName, _prepareID);
					}
				}
			}
			catch (Exception e) {}	//例外は無視する

			_prepareID = 0;
		}
	}

	/**
	 * データベース名を得る。
	 *
	 * @return	データベース名
	 */
	public String getDatabaseName()
	{
		return _databaseName;
	}

	/**
	 * プリペアステートメントIDを得る。
	 *
	 * @return	プリペアステートメントID
	 */
	public int getPrepareID()
	{
		return _prepareID;
	}
}

//
// Copyright (c) 2002, 2003, 2004, 2023 Ricoh Company, Ltd.
// All rights reserved.
//

File: sydney/Kernel/Client2/PrepareStatement.cpp
// -*-Mode: C++; tab-width: 4; c-basic-offset: 4;-*-
// vi:set ts=4 sw=4:
//
// PrepareStatement.cpp -- プリペアステートメントクラスの関数定義
// 
// Copyright (c) 2006, 2015, 2023 Ricoh Company, Ltd.
// 
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// 
//     http://www.apache.org/licenses/LICENSE-2.0
// 
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// 

namespace {
const char srcFile[] = __FILE__;
const char moduleName[] = "Client2";
}

#include "SyDefault.h"
#include "SyReinterpretCast.h"
#include "Client2/PrepareStatement.h"
#include "Client2/DataSource.h"
#include "Client2/Connection.h"
#include "Client2/Session.h"

_TRMEISTER_USING
_TRMEISTER_CLIENT2_USING

//	FUNCTION public
//	Client2::PrepareStatement::PrepareStatement -- コンストラクタ
//
//	NOTES
//
//	ARGUMENTS
//	DataSource&				cDataSource_
//		データソース
//	const ModUnicodeString&	cstrDatabaseName_
//		データベース名
//	int						iPrepareID_
//		プリペア ID
//
//	RETURN
//	なし
//
//	EXCEPTIONS

PrepareStatement::PrepareStatement(DataSource&				cDataSource_,
								   const ModUnicodeString&	cstrDatabaseName_,
								   int						iPrepareID_)
	: Object(Object::Type::PrepareStatement),
	  m_pDataSource(&cDataSource_),
	  m_pSession(0),
	  m_iPrepareID(iPrepareID_)
{
	// データベース名を代入
	m_cstrDatabaseName = cstrDatabaseName_;
}

PrepareStatement::PrepareStatement(Session*	pSession_,
								   int		iPrepareID_)
	: Object(Object::Type::PrepareStatement),
	  m_pDataSource(0),
	  m_pSession(pSession_),
	  m_iPrepareID(iPrepareID_)
{
}

//	FUNCTION public
//	Client2::PrepareStatement::~PrepareStatement -- デストラクタ
//
//	NOTES
//
//	ARGUMENTS
//	なし
//
//	RETURN
//	なし
//
//	EXCEPTIONS

PrepareStatement::~PrepareStatement()
{
	close();
}

//	FUNCTION public
//	Client2::PrepareStatement::close -- クローズする
//
//	NOTES
//
//	ARGUMENTS
//	なし
//
//	RETURN
//	なし
//
//	EXCEPTIONS

void
PrepareStatement::close()
{
	if (m_iPrepareID != 0) {

		try {
			
			if (m_pSession != 0) {

				m_pSession->erasePrepareStatement(m_iPrepareID);

			} else {

				Connection*	pClientConnection = m_pDataSource->getClientConnection();
				if (pClientConnection != 0) {

					// プリペアステートメントを削除する
					pClientConnection->erasePrepareStatement(m_cstrDatabaseName, m_iPrepareID);
				}
			}

		} catch (...) {
			Common::Thread::resetErrorCondition();
			// 例外は無視する
		}

		m_iPrepareID = 0;
	}
}

//	FUNCTION public
//	Client2::PrepareStatement::getDatabaseName -- データベース名を得る
//
//	NOTES
//
//	ARGUMENTS
//	なし
//
//	RETURN
//	const ModUnicodeString&
//		データベース名
//
//	EXCEPTIONS

const ModUnicodeString&
PrepareStatement::getDatabaseName() const
{
	return m_cstrDatabaseName;
}

//	FUNCTION public
//	Client2::PrepareStatement::getPrepareID -- プリペアステートメント ID を得る
//
//	NOTES
//
//	ARGUMENTS
//	なし
//
//	RETURN
//	int
//		プリペアステートメント ID
//
//	EXCEPTIONS

int
PrepareStatement::getPrepareID() const
{
	return m_iPrepareID;
}

//
//	Copyright (c) 2006, 2015, 2023 Ricoh Company, Ltd.
//	All rights reserved.
//
